components/OpacityMaskView/src/OpacityMaskView.cs

```diff
@@ -0,0 +1,107 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Numerics;
+#if WINDOWS_WINAPPSDK
+using Microsoft.UI.Composition;
+using Microsoft.UI.Xaml.Hosting;
+using Microsoft.UI.Xaml.Shapes;
+#else
+using Windows.UI.Composition;
+using Windows.UI.Xaml.Hosting;
+using Windows.UI.Xaml.Shapes;
+#endif
+
+namespace CommunityToolkit.WinUI.Controls;
+
+/// <summary>
+/// TODO
+/// </summary>
+[TemplatePart(Name = RootGridTemplateName, Type = typeof(Grid))]
+[TemplatePart(Name = MaskRectangleTemplateName, Type = typeof(Rectangle))]
+[TemplatePart(Name = ContentPresenterTemplateName, Type = typeof(ContentPresenter))]
+public partial class OpacityMaskView : ContentControl
+{
+    /// <summary>
+    /// Identifies the <see cref="OpacityMask"/> property.
+    /// </summary>
+    public static readonly DependencyProperty OpacityMaskProperty =
+        DependencyProperty.Register(nameof(OpacityMask), typeof(Brush), typeof(OpacityMaskView), new PropertyMetadata(null, OnOpacityMaskChanged));
+
+    private const string ContentPresenterTemplateName = "PART_ContentPresenter";
+    private const string MaskRectangleTemplateName = "PART_MaskRectangle";
+    private const string RootGridTemplateName = "PART_RootGrid";
+
+    private readonly Compositor _compositor = Window.Current.Compositor;
+    private CompositionBrush? _mask;
+    private CompositionMaskBrush? _maskBrush;
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="OpacityMaskView"/> class.
+    /// </summary>
+    public OpacityMaskView()
+    {
+        DefaultStyleKey = typeof(OpacityMaskView);
+    }
+
+    /// <summary>
+    /// Gets or sets an opacity mask, as a <see cref="Brush"/> implementation that is applied to any alpha-channel masking for the rendered content of the content.
+    /// </summary>
+    public Brush? OpacityMask
+    {
+        get => (Brush?)GetValue(OpacityMaskProperty);
+        set => SetValue(OpacityMaskProperty, value);
+    }
+
+    /// <inheritdoc />
+    protected override void OnApplyTemplate()
+    {
+        base.OnApplyTemplate();
+
+        Grid rootGrid = (Grid)GetTemplateChild(RootGridTemplateName);
+        ContentPresenter contentPresenter = (ContentPresenter)GetTemplateChild(ContentPresenterTemplateName);
+        Rectangle maskRectangle = (Rectangle)GetTemplateChild(MaskRectangleTemplateName);
+
+        _maskBrush = _compositor.CreateMaskBrush();
+        _maskBrush.Source = GetVisualBrush(contentPresenter);
+        _mask = GetVisualBrush(maskRectangle);
+        _maskBrush.Mask = OpacityMask is null ? null : _mask;
+
+        SpriteVisual redirectVisual = _compositor.CreateSpriteVisual();
+        redirectVisual.RelativeSizeAdjustment = Vector2.One;
+        redirectVisual.Brush = _maskBrush;
+        ElementCompositionPreview.SetElementChildVisual(rootGrid, redirectVisual);
```

Comment:
Would it be possible to apply any of this like we do for the `AttachedCardShadow`/`AttachedShadow` or `UIElementExtensions.VisualFactory` components in the Toolkit, so it's just an extension that can be applied to any element directly vs. a ContentControl?

Reply:
@michael-hawker I don't think we can do this. Sergio created one like `UIElementExtensions.OpacityMask` in store, but the position of the `redirectVisual` is not aligned with the source control at some times. When this happens, we use a `Border` to wrap the content to fix. I think the one in the toolkit should be more generic.

Reply:
Sounds similar to what we have to do for the `AttachedDropShadow` at times. I think having specific requirements can be fine, it's just extensions are a lot easier to add into a design than having to add another layer of depth to the Visual Tree to all cases and setup the layout. It's probably a case of a templated control vs. a raw-element with the CardShadow too I'd imagine. What are your thoughts @Sergio0694?